### PR TITLE
Reacting to Routing changes

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
Follow up for https://github.com/aspnet/Routing/pull/273

cc @rynowak

File: src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs [hunk 1]
```diff
@@ -142,15 +138,59 @@ protected virtual string GeneratePathFromRoute(string routeName, RouteValueDicti
...
+            if (!pathBase.HasValue)
             {
-                return "/";
+                if (pathData.VirtualPath.Length == 0)
+                {
+                    return "/";
+                }
+                else
+                {
+                    return !pathData.VirtualPath.StartsWith("/", StringComparison.Ordinal)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs]: `else if` YO

File: src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs [hunk 2]
```diff
@@ -142,15 +138,59 @@ protected virtual string GeneratePathFromRoute(string routeName, RouteValueDicti
...
-                return fullPath;
+                if (pathData.VirtualPath.Length == 0)
+                {
+                    return pathBase;
+                }
+                else
+                {
+                    var requiresLeadingSlash = !pathData.VirtualPath.StartsWith("/", StringComparison.Ordinal);
+                    var virtualPathStartIndex = 0;
+                    if (pathBase.Value.EndsWith("/", StringComparison.Ordinal))
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs]: Can this happen?

File: src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs [hunk 3]
```diff
@@ -142,15 +138,59 @@ protected virtual string GeneratePathFromRoute(string routeName, RouteValueDicti
...
+
+                        requiresLeadingSlash = false;
+                    }
+
+                    var virtualPathLength = pathData.VirtualPath.Length - virtualPathStartIndex;
+
+                    var builder = new StringBuilder(
+                        pathBase.Value.Length
+                        + (requiresLeadingSlash ? 1 : 0)
+                        + virtualPathLength);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs]: If you're going to always use a stringbuilder here, you can just append pathbase, trim one character if needed, and then append the virtual path. The code would be much simpler.

**Comment by rynowak** [comment]
:watch: good cleanup

File: src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs]: `PathString.Add` checked for this condition - https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Abstractions/PathString.cs#L192-L193

**Comment by pranavkm** [comment]
:up: :date:

**Comment by rynowak** [comment]
:shipit:

**Pull request closed without merging**